File: services/rag-pipeline/app/core/metrics.py
"""Prometheus metrics registry and instrumentation for RAG pipeline (Milestone 5.1).

Exports standard OpenMetrics / Prometheus text at /metrics.
"""

from __future__ import annotations

import threading
from collections.abc import Mapping
from typing import ClassVar

Labels = Mapping[str, object]


def _format_labels(labels: Labels | None) -> str:
    if not labels:
        return ""
    pairs = [
        f'{k}="{str(v).replace(chr(92), chr(92)+chr(92)).replace(chr(34), chr(92)+chr(34))}"'
        for k, v in sorted(labels.items())
        if v is not None
    ]
    return f"{{{','.join(pairs)}}}" if pairs else ""


def _labels_key(labels: Labels | None) -> str:
    if not labels:
        return ""
    return ",".join(f"{k}={v}" for k, v in sorted(labels.items()))


class Counter:
    def __init__(self, name: str, help_text: str) -> None:
        self.name = name
        self.help_text = help_text
        self._lock = threading.Lock()
        self._values: dict[str, tuple[float, Labels | None]] = {}

    def inc(self, labels: Labels | None = None, value: float = 1.0) -> None:
        if value < 0:
            raise ValueError("Counter value must be non-negative")
        key = _labels_key(labels)
        with self._lock:
            current, _ = self._values.get(key, (0.0, labels))
            self._values[key] = (current + value, labels)

    def get(self, labels: Labels | None = None) -> float:
        key = _labels_key(labels)
        with self._lock:
            val, _ = self._values.get(key, (0.0, None))
            return val

    def to_prometheus(self) -> str:
        lines = [
            f"# HELP {self.name} {self.help_text}",
            f"# TYPE {self.name} counter",
        ]
        with self._lock:
            if not self._values:
                lines.append(f"{self.name} 0")
            else:
                for val, labels in self._values.values():
                    int_or_float = int(val) if val.is_integer() else val
                    lines.append(f"{self.name}{_format_labels(labels)} {int_or_float}")
        return "\n".join(lines)


class Gauge:
    def __init__(self, name: str, help_text: str) -> None:
        self.name = name
        self.help_text = help_text
        self._lock = threading.Lock()
        self._values: dict[str, tuple[float, Labels | None]] = {}

    def set(self, value: float, labels: Labels | None = None) -> None:
        key = _labels_key(labels)
        with self._lock:
            self._values[key] = (value, labels)

    def inc(self, labels: Labels | None = None, value: float = 1.0) -> None:
        key = _labels_key(labels)
        with self._lock:
            current, _ = self._values.get(key, (0.0, labels))
            self._values[key] = (current + value, labels)

    def dec(self, labels: Labels | None = None, value: float = 1.0) -> None:
        key = _labels_key(labels)
        with self._lock:
            current, _ = self._values.get(key, (0.0, labels))
            self._values[key] = (current - value, labels)

    def get(self, labels: Labels | None = None) -> float:
        key = _labels_key(labels)
        with self._lock:
            val, _ = self._values.get(key, (0.0, None))
            return val

    def to_prometheus(self) -> str:
        lines = [
            f"# HELP {self.name} {self.help_text}",
            f"# TYPE {self.name} gauge",
        ]
        with self._lock:
            if not self._values:
                lines.append(f"{self.name} 0")
            else:
                for val, labels in self._values.values():
                    int_or_float = int(val) if val.is_integer() else val
                    lines.append(f"{self.name}{_format_labels(labels)} {int_or_float}")
        return "\n".join(lines)


DEFAULT_BUCKETS = (0.005, 0.01, 0.025, 0.05, 0.1, 0.25, 0.5, 1.0, 2.0, 5.0, 10.0)


class Histogram:
    def __init__(
        self,
        name: str,
        help_text: str,
        buckets: tuple[float, ...] = DEFAULT_BUCKETS,
    ) -> None:
        self.name = name
        self.help_text = help_text
        self.buckets = tuple(sorted(buckets))
        self._lock = threading.Lock()
        self._states: dict[str, dict[str, object]] = {}

    def observe(self, value: float, labels: Labels | None = None) -> None:
        key = _labels_key(labels)
        with self._lock:
            state = self._states.get(key)
            if state is None:
                state = {
                    "count": 0,
                    "sum": 0.0,
                    "buckets": [0] * len(self.buckets),
                    "labels": labels,
                }
                self._states[key] = state

            state["count"] = int(state["count"]) + 1  # type: ignore[index]
            state["sum"] = float(state["sum"]) + value  # type: ignore[index]
            bucket_counts = state["buckets"]
            assert isinstance(bucket_counts, list)
            for i, bound in enumerate(self.buckets):
                if value <= bound:
                    bucket_counts[i] += 1

    def to_prometheus(self) -> str:
        lines = [
            f"# HELP {self.name} {self.help_text}",
            f"# TYPE {self.name} histogram",
        ]
        with self._lock:
            if not self._states:
                for b in self.buckets:
                    lines.append(f'{self.name}_bucket{{le="{b}"}} 0')
                lines.append(f'{self.name}_bucket{{le="+Inf"}} 0')
                lines.append(f"{self.name}_sum 0")
                lines.append(f"{self.name}_count 0")
            else:
                for state in self._states.values():
                    base_labels = state["labels"] if isinstance(state["labels"], Mapping) else {}
                    bucket_counts = state["buckets"]
                    assert isinstance(bucket_counts, list)
                    for i, bound in enumerate(self.buckets):
                        lbls = {**base_labels, "le": bound}
                        lines.append(f"{self.name}_bucket{_format_labels(lbls)} {bucket_counts[i]}")
                    inf_lbls = {**base_labels, "le": "+Inf"}
                    lines.append(f"{self.name}_bucket{_format_labels(inf_lbls)} {state['count']}")
                    lines.append(f"{self.name}_sum{_format_labels(base_labels)} {state['sum']}")
                    lines.append(f"{self.name}_count{_format_labels(base_labels)} {state['count']}")
        return "\n".join(lines)


class MetricsRegistry:
    _instance: ClassVar[MetricsRegistry | None] = None

    def __init__(self) -> None:
        self.counters: dict[str, Counter] = {}
        self.gauges: dict[str, Gauge] = {}
        self.histograms: dict[str, Histogram] = {}
        self._lock = threading.Lock()

    def register_counter(self, name: str, help_text: str) -> Counter:
        with self._lock:
            if name not in self.counters:
                self.counters[name] = Counter(name, help_text)
            return self.counters[name]

    def register_gauge(self, name: str, help_text: str) -> Gauge:
        with self._lock:
            if name not in self.gauges:
                self.gauges[name] = Gauge(name, help_text)
            return self.gauges[name]

    def register_histogram(
        self,
        name: str,
        help_text: str,
        buckets: tuple[float, ...] = DEFAULT_BUCKETS,
    ) -> Histogram:
        with self._lock:
            if name not in self.histograms:
                self.histograms[name] = Histogram(name, help_text, buckets)
            return self.histograms[name]

    def get_metrics_text(self) -> str:
        parts: list[str] = []
        with self._lock:
            for c in self.counters.values():
                parts.append(c.to_prometheus())
            for g in self.gauges.values():
                parts.append(g.to_prometheus())
            for h in self.histograms.values():
                parts.append(h.to_prometheus())
        return "\n\n".join(parts) + "\n"

    def clear(self) -> None:
        with self._lock:
            self.counters.clear()
            self.gauges.clear()
            self.histograms.clear()


registry = MetricsRegistry()

# Shared platform metrics
http_server_request_duration_seconds = registry.register_histogram(
    "http_server_request_duration_seconds",
    "HTTP server request duration in seconds",
)
http_server_requests_total = registry.register_counter(
    "http_server_requests_total",
    "Total count of HTTP server requests",
)
http_server_requests_errors_total = registry.register_counter(
    "http_server_requests_errors_total",
    "Total count of HTTP server 5xx errors",
)
axiom_guardrail_pii_redactions_total = registry.register_counter(
    "axiom_guardrail_pii_redactions_total",
    "Total number of PII redactions performed during ingestion",
)


def record_request(method: str, route: str, status_code: int, duration_seconds: float) -> None:
    labels = {
        "job": "rag-pipeline",
        "method": method.upper(),
        "route": route or "unknown",
        "status": str(status_code),
    }
    http_server_request_duration_seconds.observe(duration_seconds, labels)
    http_server_requests_total.inc(labels)
    if status_code >= 500:
        http_server_requests_errors_total.inc({
            "job": "rag-pipeline",
            "method": method.upper(),
            "route": route or "unknown",
        })

File: packages/core-shared/src/metrics.ts
/**
 * Prometheus / OpenMetrics registry and instrumentation primitives.
 * Shared across TypeScript services to export standard /metrics endpoints
 * consumed by Prometheus and Grafana (O5, Milestone 5.1).
 */

export interface MetricLabels {
  [key: string]: string | number;
}

function formatLabels(labels?: MetricLabels): string {
  if (!labels) return "";
  const entries = Object.entries(labels).filter(([, v]) => v !== undefined && v !== null);
  if (entries.length === 0) return "";
  const formatted = entries
    .map(([k, v]) => `${k}="${String(v).replace(/\\/g, "\\\\").replace(/"/g, '\\"').replace(/\n/g, "\\n")}"`)
    .join(",");
  return `{${formatted}}`;
}

function labelsKey(labels?: MetricLabels): string {
  if (!labels) return "";
  return Object.entries(labels)
    .sort(([a], [b]) => a.localeCompare(b))
    .map(([k, v]) => `${k}=${v}`)
    .join(",");
}

export class Counter {
  private values = new Map<string, { value: number; labels?: MetricLabels }>();

  constructor(
    public readonly name: string,
    public readonly help: string,
  ) {}

  inc(labels?: MetricLabels, value = 1): void {
    if (value < 0) throw new Error("Counter value must be non-negative");
    const key = labelsKey(labels);
    const current = this.values.get(key)?.value ?? 0;
    this.values.set(key, { value: current + value, labels });
  }

  get(labels?: MetricLabels): number {
    return this.values.get(labelsKey(labels))?.value ?? 0;
  }

  reset(): void {
    this.values.clear();
  }

  toPrometheus(): string {
    const lines: string[] = [
      `# HELP ${this.name} ${this.help}`,
      `# TYPE ${this.name} counter`,
    ];
    if (this.values.size === 0) {
      lines.push(`${this.name} 0`);
    } else {
      for (const { value, labels } of this.values.values()) {
        lines.push(`${this.name}${formatLabels(labels)} ${value}`);
      }
    }
    return lines.join("\n");
  }
}

export class Gauge {
  private values = new Map<string, { value: number; labels?: MetricLabels }>();

  constructor(
    public readonly name: string,
    public readonly help: string,
  ) {}

  set(value: number, labels?: MetricLabels): void {
    const key = labelsKey(labels);
    this.values.set(key, { value, labels });
  }

  inc(labels?: MetricLabels, value = 1): void {
    const key = labelsKey(labels);
    const current = this.values.get(key)?.value ?? 0;
    this.values.set(key, { value: current + value, labels });
  }

  dec(labels?: MetricLabels, value = 1): void {
    const key = labelsKey(labels);
    const current = this.values.get(key)?.value ?? 0;
    this.values.set(key, { value: current - value, labels });
  }

  get(labels?: MetricLabels): number {
    return this.values.get(labelsKey(labels))?.value ?? 0;
  }

  reset(): void {
    this.values.clear();
  }

  toPrometheus(): string {
    const lines: string[] = [
      `# HELP ${this.name} ${this.help}`,
      `# TYPE ${this.name} gauge`,
    ];
    if (this.values.size === 0) {
      lines.push(`${this.name} 0`);
    } else {
      for (const { value, labels } of this.values.values()) {
        lines.push(`${this.name}${formatLabels(labels)} ${value}`);
      }
    }
    return lines.join("\n");
  }
}

export const DEFAULT_LATENCY_BUCKETS = [
  0.005, 0.01, 0.025, 0.05, 0.1, 0.25, 0.5, 1, 2, 5, 10,
];

interface HistogramBucketState {
  count: number;
  sum: number;
  buckets: number[];
  labels?: MetricLabels;
}

export class Histogram {
  private values = new Map<string, HistogramBucketState>();
  public readonly buckets: readonly number[];

  constructor(
    public readonly name: string,
    public readonly help: string,
    buckets: readonly number[] = DEFAULT_LATENCY_BUCKETS,
  ) {
    this.buckets = [...buckets].sort((a, b) => a - b);
  }

  observe(value: number, labels?: MetricLabels): void {
    const key = labelsKey(labels);
    let state = this.values.get(key);
    if (!state) {
      state = {
        count: 0,
        sum: 0,
        buckets: new Array(this.buckets.length).fill(0),
        labels,
      };
      this.values.set(key, state);
    }
    state.count += 1;
    state.sum += value;
    for (let i = 0; i < this.buckets.length; i++) {
      const bound = this.buckets[i];
      if (bound !== undefined && value <= bound) {
        state.buckets[i] = (state.buckets[i] ?? 0) + 1;
      }
    }
  }

  reset(): void {
    this.values.clear();
  }

  toPrometheus(): string {
    const lines: string[] = [
      `# HELP ${this.name} ${this.help}`,
      `# TYPE ${this.name} histogram`,
    ];
    if (this.values.size === 0) {
      for (const b of this.buckets) {
        lines.push(`${this.name}_bucket{le="${b}"} 0`);
      }
      lines.push(`${this.name}_bucket{le="+Inf"} 0`);
      lines.push(`${this.name}_sum 0`);
      lines.push(`${this.name}_count 0`);
    } else {
      for (const state of this.values.values()) {
        const baseLabels = state.labels ?? {};
        for (let i = 0; i < this.buckets.length; i++) {
          const bound = this.buckets[i];
          const count = state.buckets[i] ?? 0;
          if (bound !== undefined) {
            lines.push(
              `${this.name}_bucket${formatLabels({ ...baseLabels, le: bound })} ${count}`,
            );
          }
        }
        lines.push(
          `${this.name}_bucket${formatLabels({ ...baseLabels, le: "+Inf" })} ${state.count}`,
        );
        lines.push(
          `${this.name}_sum${formatLabels(baseLabels)} ${state.sum}`,
        );
        lines.push(
          `${this.name}_count${formatLabels(baseLabels)} ${state.count}`,
        );
      }
    }
    return lines.join("\n");
  }
}

export class MetricsRegistry {
  private counters = new Map<string, Counter>();
  private gauges = new Map<string, Gauge>();
  private histograms = new Map<string, Histogram>();

  registerCounter(name: string, help: string): Counter {
    let counter = this.counters.get(name);
    if (!counter) {
      counter = new Counter(name, help);
      this.counters.set(name, counter);
    }
    return counter;
  }

  registerGauge(name: string, help: string): Gauge {
    let gauge = this.gauges.get(name);
    if (!gauge) {
      gauge = new Gauge(name, help);
      this.gauges.set(name, gauge);
    }
    return gauge;
  }

  registerHistogram(
    name: string,
    help: string,
    buckets: readonly number[] = DEFAULT_LATENCY_BUCKETS,
  ): Histogram {
    let histogram = this.histograms.get(name);
    if (!histogram) {
      histogram = new Histogram(name, help, buckets);
      this.histograms.set(name, histogram);
    }
    return histogram;
  }

  getMetricsAsText(): string {
    const parts: string[] = [];
    for (const c of this.counters.values()) {
      parts.push(c.toPrometheus());
    }
    for (const g of this.gauges.values()) {
      parts.push(g.toPrometheus());
    }
    for (const h of this.histograms.values()) {
      parts.push(h.toPrometheus());
    }
    return parts.join("\n\n") + "\n";
  }

  clear(): void {
    this.counters.clear();
    this.gauges.clear();
    this.histograms.clear();
  }
}

export const globalMetrics = new MetricsRegistry();

export interface HttpMetrics {
  durationHistogram: Histogram;
  requestsTotal: Counter;
  errorsTotal: Counter;
  record(method: string, route: string, statusCode: number, durationSeconds: number): void;
}

export function createHttpMetrics(
  serviceName: string,
  registry: MetricsRegistry = globalMetrics,
): HttpMetrics {
  const durationHistogram = registry.registerHistogram(
    "http_server_request_duration_seconds",
    "HTTP server request duration in seconds",
  );
  const requestsTotal = registry.registerCounter(
    "http_server_requests_total",
    "Total count of HTTP server requests",
  );
  const errorsTotal = registry.registerCounter(
    "http_server_requests_errors_total",
    "Total count of HTTP server 5xx errors",
  );

  return {
    durationHistogram,
    requestsTotal,
    errorsTotal,
    record(method: string, route: string, statusCode: number, durationSeconds: number): void {
      const labels = {
        job: serviceName,
        method: method.toUpperCase(),
        route: route || "unknown",
        status: String(statusCode),
      };
      durationHistogram.observe(durationSeconds, labels);
      requestsTotal.inc(labels);
      if (statusCode >= 500) {
        errorsTotal.inc({
          job: serviceName,
          method: method.toUpperCase(),
          route: route || "unknown",
        });
      }
    },
  };
}
